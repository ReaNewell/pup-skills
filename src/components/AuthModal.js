import React from 'react';
import { connect } from 'react-redux';
import { startLogin, startLoginWithEmail, startSignUp } from '../actions/auth';

class AuthModal extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            email: "",
            error: null,
            password: ""
        }
    }
    onEmailChange = (e) => {
        const email = e.target.value;
        this.setState(() => ({ email }));
    }
    onPasswordChange = (e) => {
        const password = e.target.value;
        this.setState(() => ({ password }));
    }
    loginWithEmail = () => {
        if (this.state.email && this.state.password) {
            this.props.startLoginWithEmail(this.state.email, this.state.password);
        } else {
            this.setState(() => ({ error: "The email and/or password is invalid." }))
        }
    }
    signupWithEmail = () => {
        if (this.state.email && /^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,}$/.test(this.state.password)) {
            this.props.startSignUp(this.state.email, this.state.password);
        } else {
            this.setState(() => ({ error: "The email and/or password is invalid. Password must contain eight characters, at least one number, and at least one letter." }))
        }
    }
    render() {
        return (
            <div className="auth-modal">
                <div className="auth-modal__box">
                    {this.state.error && <p className='auth-modal__error'>{this.state.error}</p>}
                    <p className='auth-modal__exit' onClick={this.props.closeLoginModal}>Nevermind.</p>
                    <input 
                        className="auth-modal__email"
                        onChange={this.onEmailChange}
                        placeholder='Email'
                        type='email'
                        value={this.state.email}
                    />
                    <input 
                        className="auth-modal__password"
                        onChange={this.onPasswordChange}
                        placeholder='Password'
                        type='password'
                        value={this.state.password}
                    />
                    <button onClick={this.signupWithEmail} className="auth-modal__button--signUp">Sign Up</button>
                    <button onClick={this.loginWithEmail} className="auth-modal__button--signIn">Login</button>
                </div>
            </div>
        )
    };
};

const mapStateToProps = (state) => ({
    isAuthenticated: !!state.auth.uid
})

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin()),
    startSignUp: (email, password) => dispatch(startSignUp(email, password)),
    startLoginWithEmail: (email, password) => dispatch(startLoginWithEmail(email, password))
});

export default connect(mapStateToProps, mapDispatchToProps)(AuthModal);