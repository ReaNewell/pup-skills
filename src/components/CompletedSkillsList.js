import React from 'react';
import selectSkills from '../selectors/skillsByCategory';


import SkillCard from './SkillCard';

export default class CompletedSkillsList extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            completedSkills: props.activeDog.skills ? selectSkills(props.activeDog.skills, "Completed") : []
        };
    };
    componentWillReceiveProps(nextProps) {
        this.setState({ completedSkills: selectSkills(nextProps.activeDog.skills, "Completed") });
    };
    render() {
        return (
            <div className={this.props.listIsActive ? "skills-list" : "skills-list--inactive"}>
                <p className="dashboard-skills__list-title">Completed</p>
                {   
                    this.state.completedSkills.length === 0 ? (
                        <p className="skills-list__message">There are no completed skills.</p>
                    ) : (
                        this.state.completedSkills.map((skill, index) => (
                                <SkillCard
                                    key = {skill.id}
                                    skill = {skill}
                                />
                        ))
                    )
                }
            </div>
        )
    }
};