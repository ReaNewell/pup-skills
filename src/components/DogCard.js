import React from 'react';
import { connect } from 'react-redux';
import selectSkills from '../selectors/skillsByCategory';
import { activateDog } from '../actions/dogs';

export class DogCard extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            id: props.dog.id,
            name: props.dog.name,
            skills: selectSkills(props.dog.skills, "Completed").length
        }
    };
    componentWillReceiveProps(nextProps) {
        this.setState({ skills: selectSkills(nextProps.dog.skills, "Completed").length });  
    };
    // Deactivates all dogs, activates Dog Card that is clicked.
    activateDog = () => {
        this.props.activateDog(this.state.id);
        this.props.toggleBar && this.props.toggleBar();
    };
    render () {
        return (
            <div className={this.props.dog.isActive ? "dog-card dog-card--active" : "dog-card"} onClick={this.activateDog}>
                <h3>{this.state.name} knows {this.state.skills} skills.</h3>
            </div>
        );
    };
};

const mapDispatchToProps = (dispatch) => ({
    activateDog: (data) => dispatch(activateDog(data))
});

export default connect(undefined, mapDispatchToProps)(DogCard);