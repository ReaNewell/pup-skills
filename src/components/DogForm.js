import React from 'react';
import { connect } from 'react-redux';

import { startAddDog } from '../actions/dogs';

class DogForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            name: '',
            breed: '',
            error: ""
        }
    };
    onBreedChange = (e) => {
        const breed = e.target.value;
        this.setState(() => ({ breed }));
    };
    onNameChange = (e) => {
        const name = e.target.value;
        this.setState(() => ({ name }));
    };
    // Adds dog object to dogs array in Firebase, then closes the modal form.
    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.name) {
            this.setState(() => ({ error: "You must name your Pup." }));
        } else {
            this.setState(() => ({ error: "" }));
            this.props.startAddDog({
                name: this.state.name,
                breed: this.state.breed
            });
            this.setState(() => ({ name: "" }));
            this.setState(() => ({ breed: "" }));
            this.props.closeModal();
        }
    };

    render() {
        return (
            <form onSubmit={this.onSubmit} className="dog-form-window">
                <div className="dog-form">
                    <div className="dog-form__exit" onClick={this.props.closeModal}>X</div>
                    <h3 className="dog-form__title">Add your Pup</h3>
                    <div className="dog-form__entry">
                        <label className="dog-form__label">Pup's Name</label>
                        <input
                            className="dog-form__text-input"
                            onChange={this.onNameChange}
                            type="text"
                            value={this.state.name}
                        />
                    </div>
                    <div className="dog-form__entry">
                        <label className="dog-form__label">Breed</label>
                        <input
                            className="dog-form__text-input"
                            onChange={this.onBreedChange}
                            type="text"
                            value={this.state.breed}
                        />
                    </div>
                    { this.state.error && <p>{this.state.error}</p>}
                    <button className="dog-form__button">Add Dog</button>
                </div>
            </form>
        );
    };
};

const mapDispatchToProps = (dispatch) => ({
    startAddDog: (dog) => dispatch(startAddDog(dog))
});

export default connect(undefined, mapDispatchToProps)(DogForm);