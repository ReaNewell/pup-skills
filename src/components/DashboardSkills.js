import React from 'react';
import { connect } from 'react-redux';
import activeDogSelector from '../selectors/activeDog';

import CompletedSkillsList from './CompletedSkillsList';
import InProgressSkillsList from './InProgressSkillsList';
import SkillCard from './SkillCard';
import SkillsForm from './SkillsForm';

class DashboardSkills extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            dogs: props.dogs,
            activeDog: props.activeDog,
            completedListIsActive: true,
            skillModalOpen: false
        }
    }
    componentWillReceiveProps(nextProps) {
        this.setState({ dogs: nextProps.dogs, });
        this.setState({ activeDog: nextProps.activeDog });
    };

    openSkillModal = () => {
        this.setState(() => ({ skillModalOpen: true }));
    }
    closeSkillModal = () => {
        this.setState(() => ({ skillModalOpen: false }));
    }
    render() {
        return (
            <div className="dashboard-skills">
                {
                    this.state.activeDog && 
                    <div className="dashboard-skills__container">
                        <div className="dashboard-skills__header">
                            <h1 className="dashboard-skills__title">Here are {this.state.activeDog.name}'s skills</h1>
                            <button className="dashboard-skills__button" onClick={this.openSkillModal}>Add Skill</button>
                        </div>
                        {this.state.skillModalOpen && <SkillsForm closeSkillModal={this.closeSkillModal} activeDog={this.state.activeDog}/>}
                        <div className="dashboard-skills__lists">
                            <CompletedSkillsList listIsActive={this.state.completedListIsActive} activeDog={this.state.activeDog}/>
                            <InProgressSkillsList listIsActive={this.state.completedListIsActive} activeDog={this.state.activeDog}/>
                        </div>
                    </div>
                }
            </div>
        );
    };
};

const mapStateToProps = (state) => {
    return { 
        dogs: state.dogs,
        activeDog: activeDogSelector(state.dogs)
    };
};
export default connect(mapStateToProps)(DashboardSkills);