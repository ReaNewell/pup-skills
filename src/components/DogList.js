import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import DogCard from './DogCard';
import DogForm from './DogForm';

class DogList extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            modalOpen: false,
            dogCount: props.dogs.length
        }
    }
    componentWillReceiveProps(nextProps) {
        this.setState(() => ({ dogCount: nextProps.dogs.length }))
    }
    closeModal = () => {
        this.setState(() => ({ modalOpen: false }));
    }
    openModal = () => {
        this.setState(() => ({ modalOpen: true }));
    }
    render() {
        return (
            <div className="dog-list">
                {
                    this.state.dogCount === 0 ? (
                        <p>Please Add A Dog</p>
                    ) : (
                        this.props.dogs.map((dog, index) => (
                            <DogCard 
                                key = {index}
                                dog = {dog}
                            />
                        ))
                    )
                }
                {
                    this.state.dogCount < 3 && <button onClick={this.openModal} className="dog-list__button">Add Dog</button>
                }
                { this.state.modalOpen && <DogForm closeModal={this.closeModal}/>}
                <Link to='/settings'>
                    <div className="dog-list__button">Edit Profile</div>
                </Link>
                
            </div>
        )
    }
};

const mapStateToProps = (state) => ({
    dogs: state.dogs
});

export default connect(mapStateToProps)(DogList);