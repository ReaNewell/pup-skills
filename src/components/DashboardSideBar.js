import React from 'react';
import { Link } from 'react-router-dom';

import DogList from './DogList';
import NameCard from './NameCard';

class DashboardSideBar extends React.Component {
    render() {
        return (
            <div className="dashboard-sidebar__container">
                <div className="dashboard-sidebar">
                    <NameCard />
                    <DogList />
                </div>
                <div>
                    <Link to='/settings'>Edit Profile</Link>
                </div>
            </div>
        )
    };
};

export default DashboardSideBar;