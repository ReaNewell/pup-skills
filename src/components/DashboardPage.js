import React from 'react';
import {connect} from 'react-redux';

import DashboardSkills from './DashboardSkills';
import DashboardSideBar from './DashboardSideBar';
import MobileDashboardSideBar from './mobile/MobileDashboardSideBar';
import MobileDashboardSkills from './mobile/MobileDashboardSkills';

export class DashboardPage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            width: window.innerWidth
        }
    };
    componentWillMount() {
        window.addEventListener('resize', this.handleWindowSizeChange);
        if (!this.props.hasProfileName) {
            this.props.history.push('/getting-started');
        }
    };
    componentWillUnmount() {
        window.removeEventListener('resize', this.handleWindowSizeChange);
    };
    handleWindowSizeChange = () => {
        this.setState({ width: window.innerWidth });
    };
    render() {
        return (
            <div className='dashboard-page__background'>
                <div className="dashboard-page content-container">
                    {this.state.width > 720 ? <DashboardSideBar /> : <MobileDashboardSideBar/>}
                    {this.state.width > 720 ? <DashboardSkills /> : <MobileDashboardSkills/>}
                </div>
            </div>
        )
    }
};

const mapStateToProps = (state) => ({
    hasProfileName: !!state.profileInfo.profileName
});
export default connect(mapStateToProps)(DashboardPage);