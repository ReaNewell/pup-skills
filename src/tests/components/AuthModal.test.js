import React from 'react';
import { shallow } from 'enzyme';
import { AuthModal } from '../../components/AuthModal';

// RENDER TESTS
test('should render the AuthModal component correctly', () => {
    const wrapper = shallow(<AuthModal />);
    expect(wrapper).toMatchSnapshot();
});

// STARTSIGNUP TESTS
test('should call startSignUp when sign up button is clicked and email/password are provided', () => {
    const startSignUp = jest.fn();
    const wrapper = shallow(<AuthModal startSignUp={startSignUp} />);
    wrapper.setState({
        email: 'example@mail.com',
        password: 'justacode1'
    });
    wrapper.find('.auth-modal__button--signUp').simulate('click');
    expect(startSignUp).toHaveBeenCalled();
});
test('should not call startSignUp when signup button is clicked and email/password are not provided', () => {
    const startSignUp = jest.fn();
    const wrapper = shallow(<AuthModal startSignUp={startSignUp} />);
    wrapper.find('.auth-modal__button--signUp').simulate('click');
    expect(startSignUp).not.toHaveBeenCalled();
});
test('should only call startSignUp if password is more than 8 characters and includes only number and letters', () => {
    const startSignUp = jest.fn();
    const wrapper = shallow(<AuthModal startSignUp={startSignUp} />);
    wrapper.setState({
        email: 'example@mail.com',
        password: 'dddd'
    });
    wrapper.find('.auth-modal__button--signUp').simulate('click');
    expect(startSignUp).not.toHaveBeenCalled();
});

//STARTLOGINWITHEMAIL TESTS
test('should call startLoginWithEmail when login button is clicked and email/password are provided', () => {
    const startLoginWithEmail = jest.fn(() => Promise.resolve());
    const wrapper = shallow(<AuthModal startLoginWithEmail={startLoginWithEmail} />);
    wrapper.setState({
        email: 'example@mail.com',
        password: 'justacode1'
    });
    wrapper.find('.auth-modal__button--signIn').simulate('click');
    expect(startLoginWithEmail).toHaveBeenCalled();
});
test('should not call startLoginWithEmail when login button is clicked and email/password are not provided', () => {
    const startLoginWithEmail = jest.fn();
    const wrapper = shallow(<AuthModal startLoginWithEmail={startLoginWithEmail} />);
    wrapper.find('.auth-modal__button--signIn').simulate('click');
    expect(startLoginWithEmail).not.toHaveBeenCalled();
});