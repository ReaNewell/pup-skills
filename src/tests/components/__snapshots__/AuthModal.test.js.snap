// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the AuthModal component correctly 1`] = `
<div
  className="auth-modal"
>
  <div
    className="auth-modal__box"
  >
    <p
      className="auth-modal__exit"
    >
      Nevermind.
    </p>
    <input
      className="auth-modal__email"
      onChange={[Function]}
      placeholder="Email"
      required={true}
      type="email"
      value=""
    />
    <input
      className="auth-modal__password"
      onChange={[Function]}
      placeholder="Password"
      required={true}
      type="password"
      value=""
    />
    <button
      className="auth-modal__button--signUp"
      onClick={[Function]}
    >
      Sign Up
    </button>
    <button
      className="auth-modal__button--signIn"
      onClick={[Function]}
    >
      Login
    </button>
  </div>
</div>
`;
